subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java

```diff
@@ -0,0 +1,133 @@
+    ) {
+        this.projectStateRegistry = projectStateRegistry;
+        this.resolveStateFactory = resolveStateFactory;
+        this.localComponentCache = localComponentCache;
+        this.localComponentProvider = localComponentProvider;
+
+        this.localProjects = new ConcurrentCache(calculatedValueContainerFactory);
+        this.foreignProjects = new ConcurrentCache(calculatedValueContainerFactory);
+    }
+
+    @Override
+    public LocalComponentGraphResolveState getLocalComponent(ProjectComponentIdentifier projectIdentifier) {
+        return getLocalComponent(projectIdentifier, projectStateRegistry.stateFor(projectIdentifier));
+    }
+
+    private LocalComponentGraphResolveState getLocalComponent(ProjectComponentIdentifier projectIdentifier, ProjectState projectState) {
+        return localProjects.computeIfAbsent(projectIdentifier, () -> localComponentCache.computeIfAbsent(projectState, localComponentProvider::getComponent));
+    }
+
+    @Override
+    public LocalComponentGraphResolveState getForeignComponent(ProjectComponentIdentifier projectIdentifier) {
+        return foreignProjects.computeIfAbsent(projectIdentifier, () -> getComponentInAnotherBuild(projectIdentifier));
+    }
+
+    public LocalComponentGraphResolveState getComponentInAnotherBuild(ProjectComponentIdentifier projectIdentifier) {
+        ProjectState projectState = projectStateRegistry.stateFor(projectIdentifier);
+        CompositeBuildParticipantBuildState buildState = (CompositeBuildParticipantBuildState) projectState.getOwner();
+        if (buildState instanceof IncludedBuildState) {
+            // Make sure the build is configured now (not do this for the root build, as we are already configuring it right now)
+            buildState.ensureProjectsConfigured();
+        }
+
+        // Get the local component, then transform it to have a foreign identifier
+        // This accesses project state.
+        LocalComponentMetadata metadata = projectState.fromMutableState(p -> {
+            LocalComponentGraphResolveState originalComponent = getLocalComponent(projectIdentifier, projectState);
+            ProjectComponentIdentifier foreignIdentifier = buildState.idToReferenceProjectFromAnotherBuild(projectIdentifier);
+            return originalComponent.copy(foreignIdentifier, originalArtifact -> {
+                // Currently need to resolve the file, so that the artifact can be used in both a script classpath and the main build. Instead, this should be resolved as required
+                File file = originalArtifact.getFile();
+                return new CompositeProjectComponentArtifactMetadata(foreignIdentifier, originalArtifact, file);
+            });
+        });
+
+        return resolveStateFactory.stateFor(metadata);
+    }
+
+    @Override
+    public void discardAll() {
+        localProjects.clear();
+        foreignProjects.clear();
+    }
+
+    private static class ConcurrentCache {
```

Comment:
💭 This name is very generic, even for an inner type.  `ConcurrentLocalComponentCache` maybe?

Reply:
I still disagree here.  If I had a `ConcurrentList` or `SerialCache` I'd expect either a mix of elements to be contained in them, or for either type to be reusable in different domains.  This `ConcurrentCache` neither allows mixed types of elements and only applies to a single situation, so I think it's worth the extra characters to note this.

Reply:
Renamed to `ConcurrentMetadataCache`